File: SQL/src/UserDAO.java
import Database.Database;
import Database.DatabaseException;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

public class UserDAO {
    /**
     * Handles all interaction
     * with the user table
     * in the database
     */
    Connection conn;

    public UserDAO(){}

    public ArrayList<String> getUsers() throws DatabaseException{
        String sql = "SELECT * FROM Users";

        ArrayList<String> userList = null;

        try {
            PreparedStatement stmt = null;
            try {
                stmt = conn.prepareStatement(sql);
                ResultSet rs = stmt.executeQuery();

                if (rs.isBeforeFirst()) {
                    userList = new ArrayList<>();
                    // loop through result set and add games to list
                    while (rs.next()) {
                        userList.add(rs.getString("Username"));
                        userList.add(rs.getString("Password"));
                    }
                }

            } finally {
                if (stmt != null) {
                    stmt.close();
                    stmt = null;
                }
            }

        } catch (SQLException e) {
            throw new DatabaseException("Error getting all the users. ",  e);
        }

        return userList;
    }

    public boolean verifyPassword(String username, String password) throws DatabaseException {
        /**
         * get the specified user from the user table
         * given the username
         */
        String sql = "SELECT Password FROM Users WHERE Username = ?";

        try{
            PreparedStatement stmt = null;
            try {
                stmt = conn.prepareStatement(sql);
                stmt.setString(1, username);
                ResultSet rs = stmt.executeQuery();

                if (rs.isBeforeFirst()){
                    String pwd = rs.getString("Password");

                    if (pwd.equals(password)){
                        return true;
                    }
                    else{
                        return false;
                    }
                }
                else{
                    System.out.println("Could not find user to verify password");
                    return false;
                }

            }
            finally {
                if (stmt != null) {
                    stmt.close();
                    stmt = null;
                }
            }

        } catch (SQLException e) {
            throw new DatabaseException("Failed to get user " + username + " to verify password. ",  e);
        }

    }
    public boolean addNewUser(String username, String password) throws DatabaseException{
        /**
         * create the user in the database
         */
        String sql = "INSERT INTO Users (Username, Password) VALUES (?,?);";
        try{
            PreparedStatement stmt = null;
            try {
                stmt = conn.prepareStatement(sql);
                stmt.setString(1, username);
                stmt.setString(2, password);

                // update in table
                stmt.executeUpdate();
                return true;
            }
            finally {
                if (stmt != null) {
                    stmt.close();
                    stmt = null;
                }
            }

        } catch (SQLException e) {
            throw new DatabaseException("Failed to create user " + username + " " + password, e);
        }
    }

    public void setConn(Connection conn) {
        this.conn = conn;
    }
}

File: MongoDB/src/UserDAO.kt
import Database.DatabaseException
import com.mongodb.MongoClient
import com.mongodb.client.MongoCollection
import com.mongodb.client.MongoDatabase
import org.bson.Document
import java.util.ArrayList

class UserDAO(val mongoClient: MongoClient) {

    fun verifyPassword(username: String?, password: String?): Boolean {
        try {
            println("Trying to verify password for $username")
            val db = mongoClient.getDatabase("TTR")
            val collection = db.getCollection("Users")

            val doc = Document("username", username)

            val thisDoc = collection.find(doc).into(ArrayList<Document>())

            return thisDoc[0]["password"] == password
        }
        catch (e: Exception) {
            println("Verifying password didn't work")
            throw DatabaseException(e.message, e)
        }
    }

    fun addNewUser(username: String?, password: String?): Boolean {
        try {
            println("Adding user: $username")
            val db: MongoDatabase = mongoClient.getDatabase("TTR")
            val collection: MongoCollection<Document> = db.getCollection("Users")
            val doc = Document("username", username)
            doc.append("password", password)

            println("Trying to find if the user is here")
            val myDoc = collection.find(doc).first()

            if (myDoc == null) {
                collection.insertOne(doc)
                return true
            }
            else {
                return false
            }
        }
        catch (e: Exception) {
            println("Adding new user failed")
            throw DatabaseException(e.message, e)
        }
    }

    fun getUsers() : ArrayList<String> {
        try {
            println("Getting the users")
            val db: MongoDatabase = mongoClient.getDatabase("TTR")
            val collection: MongoCollection<Document> = db.getCollection("Users")

            val collector = collection.find().into(ArrayList<Document>())
            val returner = ArrayList<String>()

            for (c in collector) {
                returner.add(c["username"].toString())
                returner.add(c["password"].toString())
            }

            return returner
        } catch (e: Exception) {
            println("Getting games failed")
            throw DatabaseException(e.message, e)
        }
    }
}